sympy/simplify/simplify.py

```diff
@@ -596,6 +596,17 @@ def shorter(*choices):
 
     if expr.has(Sum):
         expr = sum_simplify(expr)
```

Comment:
I suspect it would be better to move most of the code from `sum_simplify` into a new function `factor_constants_from_sum_integral` that pulls constant terms out of Sum *and* Integral rather than just Sum.

Reply:
On second thought, maybe just making a constant extraction that is called here before `sum_simplify` that would work on Sum and Integral.

Reply:
```python
def extract_constants_from_sum_integral(expr):
    si = expr.atoms(Sum, Integral)
    reps = dict()
    for i in si:
        limsym = [a[0] for a in i.args[1:]]
        e = i.args[0]
        if isinstance(e, Add):
            e = factor_terms(i)
        c, d = e.as_independent(*limsym, asAdd=False)
        if c != 1:
            reps[i] = c*i.func(d, *i.args[1:])
    return expr.xreplace(reps)
```
```
>>> f =extract_constants_from_sum_integral
>>> f(Integral(x*y,(x,1,3))+Sum(x*y,(x,1,3)))
    3            3    
    /           __    
   |            \ `   
y* |  x dx + y*  )   x
   |            /_,   
  /            x = 1  
  1                   
```

Reply:
```
In [7]: extract_constants_from_sum_integral(Integral(x*y-z,x))                                                                                                                                               
Out[7]: 
   ⌠       
-z⋅⎮ x⋅y dx
   ⌡       

```

If an independent term is in form of sum then also it is giving output by taking as multiply.

Reply:
I left out the underscore on `as_Add`...I edited it to put that in, now.

Reply:
Note, too, that Sum does not make it to this point of code. Neither should Integral if it has (or gets) an `_eval_simplify` method. And both Sum and Integral's method should call the `extract_constants...` as their first step.

Reply:
```
In [35]:  
    ...: def extract_constants_from_sum_integral(expr): 
    ...:     si = expr.atoms(Sum, Integral) 
    ...:     reps = dict() 
    ...:     for i in si: 
    ...:         limsym = [a[0] for a in i.args[1:]] 
    ...:         e = i.args[0] 
    ...:         if isinstance(e, Add): 
    ...:             e = factor_terms(e) 
    ...:         c, d = e.as_independent(*limsym, as_Add=False) 
    ...:         if c != 1: 
    ...:             reps[i] = c*i.func(d, *i.args[1:]) 
    ...:     return expr.xreplace(reps)                                                                                                                                                                      

In [36]: extract_constants_from_sum_integral(Integral(x*y-z,x))                                                                                                                                              
Out[36]: 
⌠             
⎮ (x⋅y - z) dx
⌡             

```
I think it is not working .

Reply:
This is ok. z is not a multiplicative constant so instead of breaking this into two integrals it is being left as one. The expand method would split this up, however.